schematelemetry: emit metrics and logs about invalid objects

chrisseto opened the pull request:
Short of continuously polling `crdb_internal.invalid_objects`, there was not a convenient way to monitor a cluster for descriptor corruption.

Having such an indicator would allow customers to perform preflight checks ahead of upgrades to avoid being stuck in a mixed version state. It would also allow CRL to more easily monitor cloud clusters for corruptions in the wild.

This commit updates the schematelemetry job to additionally update the `sql.schema.invalid_objects` gauge and emit logs for any encountered corruptions.

Informs: #104266
Epic: CRDB-28665

(release note was added after commit was merged)
Release note (ops change): Added a new sql.schema.invalid_objects gauge
metric. This gauge is periodically updated based on the schedule set by
the sql.schema.telemetry.recurrence cluster setting. When it is updated,
it counts the number of schema objects (tables, types, schemas, databases,
and functions) that are in an invalid state according to CockroachDB’s
internal validation checks. This metric is expected to be zero in a healthy
cluster, and if it is not, it indicates that there is a problem that must
be repaired.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108559)
<!-- Reviewable:end -->

chrisseto commented:
Couple of notes:
I'm not sure if the metrics are named or hooked up correctly.
I had a bit of work on "canning" descriptor corruptions but ultimately tabled it to make get this work over the finish line sooner.
I would have liked to get more iterations on the job run to verify that the gauge is updated as expected but the job takes about ~30s to run or it would require exporting a bunch of internal types 🤔

andyyang890 commented on pkg/sql/catalog/schematelemetry/schema_telemetry_job.go:
```diff
@@ -66,9 +100,62 @@ func (t schemaTelemetryResumer) Resume(ctx context.Context, execCtx interface{})
+
+		var ok bool
+		count := int64(0)
+		for ok, err = rows.Next(ctx); ok; ok, err = rows.Next(ctx) {
```
I feel like this for loop might be more readable like this:
```suggestion
		for {
			ok, err = rows.Next(ctx)
			if !ok {
				break
			}
```

Also, should the `err != nil` check happen before checking `ok`?

andyyang890 commented on pkg/sql/catalog/schematelemetry/schema_telemetry_test.go:
```diff
@@ -106,7 +110,73 @@ func TestSchemaTelemetryJob(t *testing.T) {
+
+	// Assert that the InvalidObjects gauge is set to the number of expected
+	// invalid object.
+	require.Equal(t, int64(9), metrics.InvalidObjects.Value())
```
Do you mind explaining in a comment why there should be 9 invalid objects?

andyyang890 reviewed (commented):
I'm not super familiar with the jobs framework, but was interested in learning more about it so I took a stab at reviewing this PR.

One related question: do you know how/when this job is created in the first place?

fqazi reviewed (changes requested):
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108559)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto)

___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZvsGVCHy1GPl974s2:-NbZvsGVCHy1GPl974s3:b-7s0vy4) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L114)):*
> ```Go
> 	maxRecords int,
> ) error {
> 	// TODO(before merge): is this the right txn type? Does this need to be run within a txn?
> ```

Yeah this looks correct
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZwIqI9NEQEVFIS0Rc:-NbZwIqI9NEQEVFIS0Rd:b-7ifexe) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L121)):*
> ```Go
> 		}
> 
> 		rows, err := txn.QueryIteratorEx(ctx, "sql-telemetry-invalid-objects", txn.KV(), sessiondata.NodeUserSessionDataOverride, `SELECT id, error FROM "".crdb_internal.invalid_objects LIMIT `+strconv.Itoa(maxRecords))
> ```

Use $1 and pass it as a parameter.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZwjfA64sLElJmgEhV:-NbZwjfA64sLElJmgEhW:b-akhto2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L179)):*
> ```Go
> 	// Ensure that a log line is emitted for each invalid object.
> 	errorRE := regexp.MustCompile("found invalid object with ID.*")
> 	entries, err := log.FetchEntriesFromFiles(0, math.MaxInt64, 1000, errorRE, log.SelectEditMode(false, false))
> ```

Do we want to take it further and validate the types of validation errors?


<!-- Sent from Reviewable.io -->

chrisseto committed:
schematelemetry: emit metrics and logs about invalid objects

Short of continuously polling `crdb_internal.invalid_objects`, there was
not a convenient way to monitor a cluster for descriptor corruption.

Having such an indicator would allow customers to perform preflight
checks ahead of upgrades to avoid being stuck in a mixed version state.
It would also allow CRL to more easily monitor cloud clusters for
corruptions in the wild.

This commit updates the schematelemetry job to additionally update the
`sql.schema.invalid_objects` gauge and emit logs for any encountered
corruptions.

Informs: #104266
Epic: CRDB-28665
Release note: None

chrisseto reviewed (commented):
TFTRs!

> do you know how/when this job is created in the first place?

It's a combo of scheduled jobs and the `schematelemetrycontroller`package.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-:-Nbp2cEm7fUXXx0hphUP:b-y60bdy)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890 and @fqazi)

___
*[`pkg/sql/catalog/schematelemetry/scheduled_job_executor.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290854091:-NboI6cyAnsNd5uwVJbN:b9sklir) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/scheduled_job_executor.go#L30)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: maybe split the renaming into its own commit
</blockquote></details>

What would be the reasoning for that OOC? It can certainly be it's own atomic commit but it would be very small and largely irrelevant without the context of the following commit. I'm happy to do so if that's the standard practice! This seems like it may be pushing towards too granular. My "canary" for determining if something is too granular is whether or not I feel it would be possible to write a commit message longer than a sentence that provides more meaning/value than a one sentence summary.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290855514:-Nboaadm8BALVF4rrdf3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L51)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit:
```suggestion
// MetricStruct implements the metric.Struct interface.
```
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZvsGVCHy1GPl974s2:-Nbob4mj6dzdkK8sBzU8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L114)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Yeah this looks correct
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 121 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZwIqI9NEQEVFIS0Rc:-NbobbpV2nkFiIcx8Npf:bssu1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L121)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Use $1 and pass it as a parameter.
</blockquote></details>

Huh, I swear that didn't work the first time around. Fixed now!
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 132 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290872463:-NbobXf95mX6XyFNJzAc:b-umqyef) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L132)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

I feel like this for loop might be more readable like this:
```suggestion
		for {
			ok, err = rows.Next(ctx)
			if !ok {
				break
			}
```

Also, should the `err != nil` check happen before checking `ok`?
</blockquote></details>

Done.

I just copied this pattern from somewhere else :sweat_smile: 
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290877505:-Nbp1MNGBzgEqK-elNUA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L175)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Do you mind explaining in a comment why there should be 9 invalid objects?
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZwjfA64sLElJmgEhV:-Nbobikz18yMJQYnCLDa:bg09a13) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L179)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Do we want to take it further and validate the types of validation errors?
</blockquote></details>

I leaning towards no. In part there's some laziness here but I feel that quantifying the various schema corruptions within this test wouldn't provide any additional confidence about this specific test and would overall make the test more brittle.

I played around with "canning" corruptions and pairing them with a matcher. It ended up being pretty nasty given the broad scope of possible corruption and didn't change this particular test too much short of asserting that `invalid_objects` works as expected. Also additional runs of the job ended up making the test take like 90s :/.

I do think there's a good argument for a follow up PR that's focused on classifying corruptions and further testing crdb_internal.invalid_objects. WDYT?


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108559)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chrisseto and @fqazi)

___
*[`pkg/sql/catalog/schematelemetry/scheduled_job_executor.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290854091:-Nc8ToKW4e_s3nZU47Hp:b-tqw0uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/scheduled_job_executor.go#L30)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

What would be the reasoning for that OOC? It can certainly be it's own atomic commit but it would be very small and largely irrelevant without the context of the following commit. I'm happy to do so if that's the standard practice! This seems like it may be pushing towards too granular. My "canary" for determining if something is too granular is whether or not I feel it would be possible to write a commit message longer than a sentence that provides more meaning/value than a one sentence summary.
</blockquote></details>

Yeah I guess it's a judgement call, I do find it's more useful when revisiting the git history that commits that rename things have their own commit since it's immediately obvious at what point it was renamed. I'd be okay if you want to keep it in the same commit.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290855514:-Nc8WKCU6CC4RHWjyrEj:bou3nhx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L51)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

nit: missing "the" after "implements"
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290877505:-Nc8V_Zv2KLX-hD-uP8q:baeg9zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L175)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Thanks! Would you mind also adding the specific test name to the comment (i.e. `TestInvalidObjects`)?


<!-- Sent from Reviewable.io -->

chrisseto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108559)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andyyang890 and @fqazi)

___
*[`pkg/sql/catalog/schematelemetry/scheduled_job_executor.go` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290854091:-NcTUbS23r9fLMSAsLyv:bkh68lx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/scheduled_job_executor.go#L30)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Yeah I guess it's a judgement call, I do find it's more useful when revisiting the git history that commits that rename things have their own commit since it's immediately obvious at what point it was renamed. I'd be okay if you want to keep it in the same commit.
</blockquote></details>

I'll opt to keep this one bundled in given that it's a private variable!
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_job.go` line 51 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290855514:-NcTQEIf4P_xljfBTjBT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_job.go#L51)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

nit: missing "the" after "implements"
</blockquote></details>

Done.
___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#gh-1290877505:-NcTQCdZA01ehtBY_ICh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L175)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Thanks! Would you mind also adding the specific test name to the comment (i.e. `TestInvalidObjects`)?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andyyang890 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-:-NcTWBsz5QtTY2hi6Qlm:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @fqazi)



<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/108559)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andyyang890 and @chrisseto)

___
*[`pkg/sql/catalog/schematelemetry/schema_telemetry_test.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/108559#-NbZwjfA64sLElJmgEhV:-Nccptl62-nmfG25z02W:b-6wchmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/9899c85119195d55f301f50ba0948e531ad2f5bc/pkg/sql/catalog/schematelemetry/schema_telemetry_test.go#L179)):*
<details><summary><i>Previously, chrisseto (Chris Seto) wrote…</i></summary><blockquote>

I leaning towards no. In part there's some laziness here but I feel that quantifying the various schema corruptions within this test wouldn't provide any additional confidence about this specific test and would overall make the test more brittle.

I played around with "canning" corruptions and pairing them with a matcher. It ended up being pretty nasty given the broad scope of possible corruption and didn't change this particular test too much short of asserting that `invalid_objects` works as expected. Also additional runs of the job ended up making the test take like 90s :/.

I do think there's a good argument for a follow up PR that's focused on classifying corruptions and further testing crdb_internal.invalid_objects. WDYT?
</blockquote></details>

That's fair


<!-- Sent from Reviewable.io -->
